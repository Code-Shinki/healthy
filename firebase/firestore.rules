rules_version = '2';
service cloud.firestore {

  // ユーザー認証
  function isAuthUser(auth, userId) {
    return auth != null && auth.uid == userId
  }

  // スキーマ検証
  function isValidUserSchema(user) {
    return user.size() == 7 // ドキュメントが7つのプロパティを持っている
      && 'name' in user && (user.name is string || user.name == null)
      && 'gender' in user && (user.gender is string || user.gender == null)
      && 'height' in user && (user.height is number || user.height == null)
      && 'weight' in user && (user.weight is number || user.weight == null)
      && 'doctor' in user && (user.doctor is string || user.doctor == null)
      && 'createdAt' in user && (user.createdAt is string || user.createdAt == null)
      && 'health' in user && user.health is list
  }

  // 値のバリデーション
  function isValidUserData(user) {
    return (user.name.size() <= 30 || user.name == null)
      && (user.gender.matches('男性|女性') || user.gender == null)
      && ((50 <= user.height && user.height <= 300) || user.height == null)
      && ((10 <= user.weight && user.weight <= 300) || user.weight == null)
      && (user.doctor.size() <= 30 || user.doctor == null)
      && (user.createdAt.size() <= 50 || user.createdAt == null)
  }

  match /databases/{database}/documents {
    match /users/{userId} {
      allow create: if isAuthUser(request.auth, userId) && isValidUserSchema(request.resource.data) && isValidUserData(request.resource.data);
      allow read:  if isAuthUser(request.auth, userId);
      allow update: if isAuthUser(request.auth, userId) && isValidUserSchema(request.resource.data) && isValidUserData(request.resource.data);
      allow delete: if isAuthUser(request.auth, userId);
    }
  }
}
