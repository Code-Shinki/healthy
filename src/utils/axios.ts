import axios from 'axios'

const instance = axios.create()

instance.interceptors.request.use((config) => {
  config.headers.common['Authorization'] = 'Bearer ' + process.env.NEXT_PUBLIC_X_API_KEY
  return config
})

export default instance
