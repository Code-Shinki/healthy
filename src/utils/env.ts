export const SITE_DOMAIN = process.env.NEXT_PUBLIC_SITE_DOMAIN || ''
export const SITE_TITLE = process.env.NEXT_PUBLIC_SITE_TITLE || ''
export const SITE_SUBTITLE = process.env.NEXT_PUBLIC_SITE_SUBTITLE || ''
export const SITE_DESCRIPTION = process.env.NEXT_PUBLIC_SITE_DESCRIPTION || ''

export const FIREBASE_API_KEY = process.env.NEXT_PUBLIC_FIREBASE_API_KEY || ''
export const FIREBASE_AUTH_DOMAIN = process.env.NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN || ''
export const FIREBASE_PROJECT_ID = process.env.NEXT_PUBLIC_FIREBASE_PROJECT_ID || ''
export const FIREBASE_STORAGE_BUCKET = process.env.NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET || ''
export const FIREBASE_MESSAGING_SENDER_ID = process.env.NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID || ''
export const FIREBASE_APP_ID = process.env.NEXT_PUBLIC_FIREBASE_APP_ID || ''
