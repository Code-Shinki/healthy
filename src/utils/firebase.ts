import firebase from 'firebase/app'
import 'firebase/auth'
import 'firebase/firestore'
import {
  FIREBASE_API_KEY,
  FIREBASE_APP_ID,
  FIREBASE_AUTH_DOMAIN,
  FIREBASE_MESSAGING_SENDER_ID,
  FIREBASE_PROJECT_ID,
  FIREBASE_STORAGE_BUCKET,
} from 'utils/env'

const config = {
  apiKey: FIREBASE_API_KEY,
  authDomain: FIREBASE_AUTH_DOMAIN,
  projectId: FIREBASE_PROJECT_ID,
  storageBucket: FIREBASE_STORAGE_BUCKET,
  messagingSenderId: FIREBASE_MESSAGING_SENDER_ID,
  appId: FIREBASE_APP_ID,
}
if (firebase.apps.length === 0) {
  firebase.initializeApp(config)
}

if (process.browser) {
  ;(async () => {
    // firestoreのキャッシュを有効化
    await firebase.firestore().enablePersistence({ synchronizeTabs: true })
  })()
}

const auth = firebase.auth()
const db = firebase.firestore()

export { auth, db }
