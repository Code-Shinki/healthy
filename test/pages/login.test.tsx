import { cleanup, fireEvent, render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { useRouter } from 'next/router'
import React from 'react'
import Login from '../../src/pages/login'
import { auth } from '../../src/utils/firebase'
import AppMock from '../mocks/app-mock'
import { loggedInDemo, loggedInInit, nowLoading, unregistered } from '../mocks/globalStateMock'
import { nextRouterMock } from '../mocks/nextRouterMock'

// モック関数を作成
jest.mock('next/router', () => ({ useRouter: jest.fn() }))
jest.mock('../../src/utils/firebase', () => ({
  auth: { signInWithEmailAndPassword: jest.fn() },
}))
jest.mock('../../src/requests/userDataset', () => ({
  getUserDataset: jest.fn(),
  postUserDataset: jest.fn(),
  deleteUserDataset: jest.fn(),
}))

describe(`ログインページ (/login)`, () => {
  afterEach(cleanup)

  test('スナップショット', async () => {
    // useRouterのメソッドをモック化する
    const mockRouter = nextRouterMock
    ;(useRouter as jest.Mock).mockReturnValue(mockRouter)

    const { asFragment } = render(
      <AppMock initializeState={unregistered} drawerMenu={false}>
        <Login />
      </AppMock>
    )

    const tree = asFragment()
    expect(tree).toMatchSnapshot()
  })

  test('ログイン状態を判別中', async () => {
    // ソースファイルのrouterに渡した引数をmockRouterで受け取る
    const mockRouter = nextRouterMock
    ;(useRouter as jest.Mock).mockReturnValue(mockRouter)

    render(
      <AppMock initializeState={nowLoading} drawerMenu={false}>
        <Login />
      </AppMock>
    )

    // 読み込み中にcomponents/atoms/spinner.tsxが表示されるか
    expect(screen.getByText('読み込んでいます')).toBeInTheDocument()
  })

  test('ログインしていないユーザー', async () => {
    const mockRouter = nextRouterMock
    ;(useRouter as jest.Mock).mockReturnValue(mockRouter)

    render(
      <AppMock initializeState={unregistered} drawerMenu={false}>
        <Login />
      </AppMock>
    )

    // メニューが存在しないか
    expect(screen.queryByRole('button', { name: 'ホーム' })).not.toBeInTheDocument()

    // メールアドレスを入力するテキストフィールドが存在し、変更できる
    const email = screen.getByLabelText('email').querySelector('input') as HTMLInputElement
    expect(email.value).toEqual('')
    fireEvent.change(email, { target: { value: 'test' } }) // テキストを変更
    expect(email.value).toEqual('test')

    // パスワードを入力するテキストフィールドが存在し、変更できる]
    const password = screen.getByLabelText('password').querySelector('input') as HTMLInputElement
    expect(password.value).toEqual('')
    fireEvent.change(password, { target: { value: 'Z9mLGk3R' } }) // テキストを変更
    expect(password.value).toEqual('Z9mLGk3R')

    // ログインボタンをクリック
    userEvent.click(screen.getByRole('button', { name: 'ログイン', hidden: true }))

    // メールアドレスのバリデートが走る
    expect(screen.getByText('有効なメールアドレスを入力してください。')).toBeInTheDocument()

    // 有効なメールアドレスに変更して登録ボタンをクリック
    fireEvent.change(email, { target: { value: 'test@test.test' } })
    userEvent.click(screen.getByRole('button', { name: 'ログイン', hidden: true }))

    // ユーザー登録のリクエストが走った
    expect(auth.signInWithEmailAndPassword).toHaveBeenCalled()
  })

  test('初期データを持つログイン済みのユーザー', async () => {
    const mockRouter = nextRouterMock
    ;(useRouter as jest.Mock).mockReturnValue(mockRouter)

    render(
      <AppMock initializeState={loggedInInit} drawerMenu={false}>
        <Login />
      </AppMock>
    )

    // next/routerでダッシュボードページへリダイレクトしているか
    expect(mockRouter.push).toHaveBeenCalledWith('/dashboard')
  })

  test('フルデータを持つログイン済みのユーザー', async () => {
    // useRouterのメソッドをモック化する
    const mockRouter = nextRouterMock
    ;(useRouter as jest.Mock).mockReturnValue(mockRouter)

    render(
      <AppMock initializeState={loggedInDemo} drawerMenu={false}>
        <Login />
      </AppMock>
    )

    // next/routerでダッシュボードページへリダイレクトしているか
    expect(mockRouter.push).toHaveBeenCalledWith('/dashboard')
  })
})
