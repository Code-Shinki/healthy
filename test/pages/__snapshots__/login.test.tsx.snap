// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ログインページ (/login) スナップショット 1`] = `
<DocumentFragment>
  <main
    class="MuiGrid-root makeStyles-root-1 MuiGrid-container"
  >
    <div
      class="MuiPaper-root MuiGrid-root makeStyles-paper-2 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-8 MuiGrid-grid-md-6 MuiGrid-grid-lg-5 MuiPaper-elevation6"
    >
      <div
        class="makeStyles-container-3"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-4 MuiAvatar-colorDefault"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            style="font-size: 2.5em;"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 17c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm6-9h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6h1.9c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm0 12H6V10h12v10z"
            />
          </svg>
        </div>
        <h1>
          ログイン
        </h1>
        <p>
          ※ 現在ログアウト時にデータを削除しているため利用できません。
        </p>
        <form
          class="makeStyles-form-6"
          novalidate=""
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <div
                aria-label="email"
                class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                  data-shrink="false"
                  for="email"
                  id="email-label"
                >
                  メールアドレス
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                  >
                     *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    autocomplete="email"
                    class="MuiInputBase-input MuiOutlinedInput-input"
                    id="email"
                    name="email"
                    required=""
                    type="email"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-8 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-10"
                    >
                      <span>
                        メールアドレス *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <div
                aria-label="password"
                class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                  data-shrink="false"
                  for="password"
                  id="password-label"
                >
                  パスワード
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                  >
                     *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    autocomplete="current-password"
                    class="MuiInputBase-input MuiOutlinedInput-input"
                    id="password"
                    name="password"
                    required=""
                    type="password"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-8 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-10"
                    >
                      <span>
                        パスワード *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-submit-7 MuiButton-outlinedPrimary MuiButton-outlinedSizeLarge MuiButton-sizeLarge MuiButton-fullWidth"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              ログイン
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </form>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-flex-end"
        >
          <a
            class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
            href="/signup"
          >
            新規作成はこちら
          </a>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root makeStyles-image-5 MuiGrid-item MuiGrid-grid-sm-4 MuiGrid-grid-md-6 MuiGrid-grid-lg-7"
    />
  </main>
</DocumentFragment>
`;
